File: common/src/test/java/wile/redstonepen/blocks/RedstoneTrackDefsTest.java
package wile.redstonepen.blocks;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;

import net.minecraft.core.Direction;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.junit.jupiter.api.Test;
import wile.redstonepen.blocks.track.RedstoneTrackDefs;
import static wile.redstonepen.blocks.track.RedstoneTrackDefs.*;

class RedstoneTrackDefsTest
{
  // --- shape.get covers every face-bit combination -------------------------------------------

  @Test
  void shapeGetReturnsNonNullForEveryFaceMask()
  {
    for(int faces=0; faces<64; ++faces) {
      final VoxelShape s = RedstoneTrackDefs.shape.get(faces);
      assertNotNull(s);
    }
  }

  @Test
  void shapeGetIsCachedForRepeatedCalls()
  {
    final VoxelShape a = RedstoneTrackDefs.shape.get(0x3F);
    final VoxelShape b = RedstoneTrackDefs.shape.get(0x3F);
    assertEquals(a, b);
  }

  @Test
  void shapeForNoFacesIsEmpty()
  {
    assertTrue(RedstoneTrackDefs.shape.get(0).isEmpty());
  }

  @Test
  void shapeForAnyFaceIsNonEmpty()
  {
    for(int bit=0; bit<6; ++bit) {
      assertFalse(RedstoneTrackDefs.shape.get(1<<bit).isEmpty());
    }
  }

  // --- models maps are populated -------------------------------------------------------------

  @Test
  void wireMappingHas25Entries()
  { assertEquals(25, RedstoneTrackDefs.models.STATE_WIRE_MAPPING.size()); }

  @Test
  void connectMappingHas7Entries()
  { assertEquals(7, RedstoneTrackDefs.models.STATE_CONNECT_MAPPING.size()); }

  // --- connections bit helpers ---------------------------------------------------------------

  @Test
  void connectionBitOrderHasSixDirections()
  {
    assertEquals(6, RedstoneTrackDefs.connections.CONNECTION_BIT_ORDER.length);
    assertEquals(6, RedstoneTrackDefs.connections.CONNECTION_BIT_ORDER_REV.size());
  }

  @Test
  void getBulkConnectorBitMatchesForwardMapping()
  {
    for(Direction d : Direction.values()) {
      final long bit = RedstoneTrackDefs.connections.getBulkConnectorBit(d);
      assertEquals(d, RedstoneTrackDefs.connections.BULK_FACE_MAPPING.get(bit));
    }
  }

  @Test
  void getWireBitYieldsKnownValues()
  {
    assertEquals(0x00000001L, RedstoneTrackDefs.connections.getWireBit(Direction.DOWN, Direction.NORTH));
    assertEquals(0x00800000L, RedstoneTrackDefs.connections.getWireBit(Direction.WEST, Direction.SOUTH));
  }

  @Test
  void getWireBitForUnknownPairReturnsZero()
  {
    assertEquals(0L, RedstoneTrackDefs.connections.getWireBit(Direction.DOWN, Direction.DOWN));
  }

  @Test
  void getWireBitSideAndDirectionRoundTrip()
  {
    final long bit = RedstoneTrackDefs.connections.getWireBit(Direction.NORTH, Direction.EAST);
    final var t = RedstoneTrackDefs.connections.getWireBitSideAndDirection(bit);
    assertEquals(Direction.NORTH, t.getA());
    assertEquals(Direction.EAST, t.getB());
  }

  @Test
  void getWireBitSideAndDirectionUnknownDefaultsDownDown()
  {
    final var t = RedstoneTrackDefs.connections.getWireBitSideAndDirection(0xFFFFFFFFL);
    assertEquals(Direction.DOWN, t.getA());
    assertEquals(Direction.DOWN, t.getB());
  }

  @Test
  void getVanillaWireConnectionDirectionsReturnsAllSetSides()
  {
    final List<Direction> dirs = RedstoneTrackDefs.connections.getVanillaWireConnectionDirections(0x0FL);
    assertEquals(4, dirs.size());
    assertTrue(dirs.contains(Direction.NORTH));
    assertTrue(dirs.contains(Direction.SOUTH));
    assertTrue(dirs.contains(Direction.EAST));
    assertTrue(dirs.contains(Direction.WEST));
  }

  @Test
  void getVanillaWireConnectionDirectionsEmptyForZeroMask()
  {
    assertTrue(RedstoneTrackDefs.connections.getVanillaWireConnectionDirections(0L).isEmpty());
  }

  @Test
  void hasVanillaWireConnectionPerSide()
  {
    assertTrue(RedstoneTrackDefs.connections.hasVanillaWireConnection(0x01, Direction.NORTH));
    assertTrue(RedstoneTrackDefs.connections.hasVanillaWireConnection(0x02, Direction.SOUTH));
    assertTrue(RedstoneTrackDefs.connections.hasVanillaWireConnection(0x04, Direction.EAST));
    assertTrue(RedstoneTrackDefs.connections.hasVanillaWireConnection(0x08, Direction.WEST));
    assertFalse(RedstoneTrackDefs.connections.hasVanillaWireConnection(0x01, Direction.SOUTH));
    assertFalse(RedstoneTrackDefs.connections.hasVanillaWireConnection(0L, Direction.UP));
  }

  @Test
  void hasBulkConnectionDetectsFaceBit()
  {
    final long mask = RedstoneTrackDefs.connections.getBulkConnectorBit(Direction.EAST);
    assertTrue(RedstoneTrackDefs.connections.hasBulkConnection(mask, Direction.EAST));
    assertFalse(RedstoneTrackDefs.connections.hasBulkConnection(mask, Direction.WEST));
  }

  @Test
  void hasRedstoneConnectionAllSidesAtLeastOneBitTrue()
  {
    for(Direction d : Direction.values()) {
      // some bit per direction makes the predicate true
      final long full = 0xFFFFFFFFL;
      assertTrue(RedstoneTrackDefs.connections.hasRedstoneConnection(full, d));
    }
  }

  @Test
  void hasRedstoneConnectionFalseForZero()
  {
    for(Direction d : Direction.values()) {
      assertFalse(RedstoneTrackDefs.connections.hasRedstoneConnection(0L, d));
    }
  }

  @Test
  void getWireElementsOnFaceCovers4BitsPerFace()
  {
    for(Direction d : Direction.values()) {
      final long mask = RedstoneTrackDefs.connections.getWireElementsOnFace(d);
      assertEquals(4L, Long.bitCount(mask));
    }
  }

  @Test
  void getAllElementsOnFaceCoversWiresAndConnector()
  {
    for(Direction d : Direction.values()) {
      final long mask = RedstoneTrackDefs.connections.getAllElementsOnFace(d);
      assertEquals(5L, Long.bitCount(mask));
    }
  }

  @Test
  void redstoneUpdateDirectionsContainsAllSixFaces()
  {
    assertEquals(6, RedstoneTrackDefs.REDSTONE_UPDATE_DIRECTIONS.length);
  }
}

File: common/src/test/kotlin/wile/redstonepen/blocks/RedstoneTrackDefsTest.kt
package wile.redstonepen.blocks

import io.kotest.assertions.assertSoftly
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldContain
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import net.minecraft.core.Direction
import wile.redstonepen.blocks.track.RedstoneTrackDefs
import wile.redstonepen.blocks.track.RedstoneTrackDefs.Connections

class RedstoneTrackDefsTest :
    DescribeSpec({
        describe("shape") {
            it("get returns non-null for every face mask") {
                for (faces in 0 until 64) {
                    RedstoneTrackDefs.Shape.get(faces) shouldNotBe null
                }
            }

            it("get is cached for repeated calls") {
                RedstoneTrackDefs.Shape.get(0x3F) shouldBe RedstoneTrackDefs.Shape.get(0x3F)
            }

            it("no faces yields empty shape") {
                RedstoneTrackDefs.Shape.get(0).isEmpty shouldBe true
            }

            it("any single face yields non-empty shape") {
                for (bit in 0 until 6) {
                    RedstoneTrackDefs.Shape.get(1 shl bit).isEmpty shouldBe false
                }
            }
        }

        describe("models") {
            it("wire mapping has 25 entries") {
                RedstoneTrackDefs.Models.STATE_WIRE_MAPPING.size shouldBe 25
            }

            it("connect mapping has 7 entries") {
                RedstoneTrackDefs.Models.STATE_CONNECT_MAPPING.size shouldBe 7
            }
        }

        describe("connections") {
            it("bit order has six directions") {
                assertSoftly {
                    Connections.CONNECTION_BIT_ORDER.size shouldBe 6
                    Connections.CONNECTION_BIT_ORDER_REV.size shouldBe 6
                }
            }

            it("bulk connector bit matches forward mapping") {
                assertSoftly {
                    for (d in Direction.values()) {
                        val bit = Connections.getBulkConnectorBit(d)
                        Connections.BULK_FACE_MAPPING[bit] shouldBe d
                    }
                }
            }

            it("getWireBit yields known values") {
                assertSoftly {
                    Connections.getWireBit(Direction.DOWN, Direction.NORTH) shouldBe 0x00000001L
                    Connections.getWireBit(Direction.WEST, Direction.SOUTH) shouldBe 0x00800000L
                }
            }

            it("getWireBit returns zero for same-direction pair") {
                Connections.getWireBit(Direction.DOWN, Direction.DOWN) shouldBe 0L
            }

            it("getWireBit side and direction round trip") {
                val bit = Connections.getWireBit(Direction.NORTH, Direction.EAST)
                val t = Connections.getWireBitSideAndDirection(bit)
                assertSoftly {
                    t.a shouldBe Direction.NORTH
                    t.b shouldBe Direction.EAST
                }
            }

            it("getWireBitSideAndDirection defaults to DOWN/DOWN for unknown bit") {
                val t = Connections.getWireBitSideAndDirection(0xFFFFFFFFL)
                assertSoftly {
                    t.a shouldBe Direction.DOWN
                    t.b shouldBe Direction.DOWN
                }
            }

            it("getVanillaWireConnectionDirections returns all set sides") {
                val dirs = Connections.getVanillaWireConnectionDirections(0x0FL)
                assertSoftly {
                    dirs.size shouldBe 4
                    dirs shouldContain Direction.NORTH
                    dirs shouldContain Direction.SOUTH
                    dirs shouldContain Direction.EAST
                    dirs shouldContain Direction.WEST
                }
            }

            it("getVanillaWireConnectionDirections returns empty for zero mask") {
                Connections.getVanillaWireConnectionDirections(0L).shouldBeEmpty()
            }

            it("hasVanillaWireConnection per side") {
                assertSoftly {
                    Connections.hasVanillaWireConnection(0x01, Direction.NORTH) shouldBe true
                    Connections.hasVanillaWireConnection(0x02, Direction.SOUTH) shouldBe true
                    Connections.hasVanillaWireConnection(0x04, Direction.EAST) shouldBe true
                    Connections.hasVanillaWireConnection(0x08, Direction.WEST) shouldBe true
                    Connections.hasVanillaWireConnection(0x01, Direction.SOUTH) shouldBe false
                    Connections.hasVanillaWireConnection(0L, Direction.UP) shouldBe false
                }
            }

            it("hasBulkConnection detects face bit") {
                val mask = Connections.getBulkConnectorBit(Direction.EAST)
                assertSoftly {
                    Connections.hasBulkConnection(mask, Direction.EAST) shouldBe true
                    Connections.hasBulkConnection(mask, Direction.WEST) shouldBe false
                }
            }

            it("hasRedstoneConnection is true for all directions when all bits set") {
                val full = 0xFFFFFFFFL
                assertSoftly {
                    for (d in Direction.values()) {
                        Connections.hasRedstoneConnection(full, d) shouldBe true
                    }
                }
            }

            it("hasRedstoneConnection is false for all directions when zero") {
                assertSoftly {
                    for (d in Direction.values()) {
                        Connections.hasRedstoneConnection(0L, d) shouldBe false
                    }
                }
            }

            it("getWireElementsOnFace covers 4 bits per face") {
                assertSoftly {
                    for (d in Direction.values()) {
                        val mask = Connections.getWireElementsOnFace(d)
                        java.lang.Long.bitCount(mask) shouldBe 4
                    }
                }
            }

            it("getAllElementsOnFace covers wires and connector") {
                assertSoftly {
                    for (d in Direction.values()) {
                        val mask = Connections.getAllElementsOnFace(d)
                        java.lang.Long.bitCount(mask) shouldBe 5
                    }
                }
            }

            it("REDSTONE_UPDATE_DIRECTIONS contains all six faces") {
                RedstoneTrackDefs.REDSTONE_UPDATE_DIRECTIONS.size shouldBe 6
            }

            it("getWireElementsOnFace masks are disjoint across all faces") {
                val masks = Direction.values().map { Connections.getWireElementsOnFace(it) }
                assertSoftly {
                    for (i in masks.indices) {
                        for (j in masks.indices) {
                            if (i != j) (masks[i] and masks[j]) shouldBe 0L
                        }
                    }
                }
            }

            it("getAllElementsOnFace masks are disjoint across all faces") {
                val masks = Direction.values().map { Connections.getAllElementsOnFace(it) }
                assertSoftly {
                    for (i in masks.indices) {
                        for (j in masks.indices) {
                            if (i != j) (masks[i] and masks[j]) shouldBe 0L
                        }
                    }
                }
            }
        }
    })
